backend/fms_core/services/report.py

```diff
@@ -0,0 +1,134 @@
+from typing import List
+import datetime
+from collections import defaultdict
+import pandas as pd
+from django.apps import apps
+from django.db.models import F, Count,  Sum, Max, Min, TextChoices, functions
+
+from fms_report.models import Report, MetricField
+from fms_report.models._constants import AggregationType
+
+class TimeWindow(TextChoices):
+    MONTHLY = "month", "Monthly"
+    WEEKLY = "week", "Weekly"
+    DAILY = "day", "Daily"
+
+def get_date_range_with_window(start_date: str, end_date: str, time_window: TimeWindow):
+    time_series = pd.date_range(start=start_date, end=end_date).to_series()
+    match time_window:
+        case TimeWindow.MONTHLY: 
+            window_time_series = time_series.to_numpy().astype('datetime64[M]')
+        case TimeWindow.WEEKLY:
+            window_time_series = time_series - pd.to_timedelta(time_series.dt.dayofweek, unit="D")
+        case TimeWindow.DAILY:
+             window_time_series = time_series
+
+    date_range = pd.to_datetime(time_series, unit='D')
+    time_window_range = pd.to_datetime(window_time_series, unit='D')
+    return date_range, time_window_range
+
+def list_reports():
+    return Report.objects.values_list("name", flat=True)
+
+def list_report_information(name: str):
+    queryset = MetricField.objects.filter(report__name=name).all()
+    groups = [field.name for field in queryset if field.is_group]
+    time_windows = [TimeWindow.DAILY.label, TimeWindow.WEEKLY.label, TimeWindow.MONTHLY.label]
+    report_info = {"name": name,
+                   "groups": groups,
+                   "time_windows": time_windows}
+    return report_info
+
+def get_report(name: str, grouped_by: List[str], time_window: TimeWindow, start_date: str, end_date: str):
```

Comment:
I'd be nice to have a `class SerializedReport(TypedDict)` so I can know the format before looking at the code ;)

Reply:
We'll tweak the format first and once it crystalizes we can start typing to your heart content :)